formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/Encoder.kt

```diff
@@ -0,0 +1,356 @@
+ */
+@file:OptIn(ExperimentalSerializationApi::class, ExperimentalUnsignedTypes::class)
+
+package kotlinx.serialization.cbor.internal
+
+import kotlinx.serialization.*
+import kotlinx.serialization.builtins.*
+import kotlinx.serialization.cbor.*
+import kotlinx.serialization.descriptors.*
+import kotlinx.serialization.encoding.*
+import kotlinx.serialization.modules.*
+import kotlin.experimental.*
+
+
+//value classes are only inlined on the JVM, so we use a typealias and extensions instead
+private typealias Stack = MutableList<CborWriter.Data>
+
+private fun Stack(vararg elements: CborWriter.Data): Stack = mutableListOf(*elements)
+private fun Stack.push(value: CborWriter.Data) = add(value)
+private fun Stack.pop() = removeLast()
+private fun Stack.peek() = last()
+
+// Writes class as map [fieldName, fieldValue]
+// Split implementation to optimize base case
+internal sealed class CborWriter(
+    protected val cbor: Cbor,
+    protected val output: ByteArrayOutput,
+) : AbstractEncoder() {
+    protected var isClass = false
+
+    protected var encodeByteArrayAsByteString = false
+
+    class Data(val bytes: ByteArrayOutput, var elementCount: Int)
+
+    protected abstract fun getDestination(): ByteArrayOutput
+
+    override val serializersModule: SerializersModule
+        get() = cbor.serializersModule
+
+
+    @OptIn(ExperimentalSerializationApi::class)
+    override fun <T> encodeSerializableValue(serializer: SerializationStrategy<T>, value: T) {
+
+        if ((encodeByteArrayAsByteString || cbor.alwaysUseByteString)
+            && serializer.descriptor == ByteArraySerializer().descriptor
+        ) {
+            getDestination().encodeByteString(value as ByteArray)
+        } else {
+            encodeByteArrayAsByteString = encodeByteArrayAsByteString || serializer.descriptor.isInlineByteString()
+            super.encodeSerializableValue(serializer, value)
+        }
+    }
+
+    override fun shouldEncodeElementDefault(descriptor: SerialDescriptor, index: Int): Boolean = cbor.encodeDefaults
+
+    override fun encodeElement(descriptor: SerialDescriptor, index: Int): Boolean {
+        isClass = descriptor.getElementDescriptor(index).kind == StructureKind.CLASS
+        encodeByteArrayAsByteString = descriptor.isByteString(index)
+
+        val name = descriptor.getElementName(index)
+        val label = descriptor.getCborLabel(index)
+
+        if (!descriptor.hasArrayTag()) {
+            if (cbor.writeKeyTags) descriptor.getKeyTags(index)?.forEach { getDestination().encodeTag(it) }
+
+            if ((descriptor.kind !is StructureKind.LIST) && (descriptor.kind !is StructureKind.MAP) && (descriptor.kind !is PolymorphicKind)) {
+                //indices are put into the name field. we don't want to write those, as it would result in double writes
+                if (cbor.preferCborLabelsOverNames && label != null) {
+                    getDestination().encodeNumber(label)
```

Comment:
Since we do not need label value unless `cbor.preferCborLabelsOverNames` is true, it's better to move `val label = descriptor.getCborLabel(index)` line under this if so we won't access annotation list unless it is necessary.